plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx

```diff
@@ -33,7 +33,7 @@ import {
 import './editor.scss';
 import { getDefaultStockStatuses } from '../product-collection/constants';
 
-const DEFAULT_QUERY_CONTEXT_ATTRIBUTES = [ 'collection' ];
+const DEFAULT_QUERY_CONTEXT_ATTRIBUTES = [ 'collection', 'selectedReference' ];
```

Comment:
This seems tricky. On one hand, having precise control over different settings makes sense, but for the sake of simplicity, it might be worth considering overriding the `location` result (from the useGetLocation) that is already included in the query request instead of adding it here.

I'm considering this from the perspective of a PC block being moved within a Single Product block. This action would change the `location` request argument, similar to what should happen when a product reference is selected. 

What are your thoughts on this? cc @kmanijak

Reply:
I'm not suggesting to relocate this logic into the `useGetLocation`. While this could be a good idea if the selected-product-reference becomes a universal way of handling context-dependent blocks, for now, I believe it's best to maintain consistency and make it easy for third-party developers to not have to worry about that difference, at least for the time being.

Reply:
This also makes me wonder whether the `location` itself should be included in the query context here. Just like it does in the frontend 🤔

Reply:
To double check if I understand correctly - you suggest a logic like:
- include `location` in query by default
- if the reference has been selected by default, override the `location`?

I totally agree we should expose SINGLE information. IMO for 3PD it's important what's the reference and it's not important if it was recognised automatically (`location`) or manually (`selectedReference`) and not both for 3PD to choose from 🤔 Is that your point Chris?

Reply:
Yes, exactly! We should streamline only one piece of information! The same handling should also happen in PHP as a follow-up on this.

Reply:
> include location in query by default

Furthermore, I was considering that the "location" can be seen as a form of "context," so it would be logical to incorporate it into the `queryContextIncludes` request parameter. I also think that both the `location` and `queryContextIncludes` could be improved with some renaming, perhaps through a separate PR.

So in summary:
- We are combining the 'location' and the 'selectedReference' in this pull request and then;
- We follow up with another pull request to refine these names so that they all pertain to the term "reference."

Reply:
I also agree to have SINGLE information 💯

**Question:** What should we call this SINGLE information? `selectedReference` or `location`? 🤔

We are only exposing `usesReference` to 3PDs. Therefore, IMO terminology should be around `usesReference`. 
- `selectedReference` is more about what is selected by merchant manually. Therefore IMO it doesn't make sense to call it `selectedReference` because sometimes it's automatically calculated i.e. `location`.
- `location` is more about where the block is being used. Therefore IMO it doesn't make sense to call it `location` because sometimes it's manually selected by merchant i.e. `selectedReference`.

I believe it will require a new name like `availableReference` or `availableContext` (😅). What do you think? 
- **Sidenote:** We are using name [productCollectionLocation](https://github.com/woocommerce/woocommerce/blob/53e1da8ee51d0ef76e4d13dd63cc287bd013d177/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L132) on Frontend, which contain calculated location. Therefore, we will need to update that name too in case we decide to combine these two.


---

> This also makes me wonder whether the location itself should be included in the query context here. Just like it is on the frontend 🤔

@Chris I think we are already doing that [here](https://github.com/woocommerce/woocommerce/blob/53e1da8ee51d0ef76e4d13dd63cc287bd013d177/plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx#L268), unless you are referring to something else. 🤔

Reply:
@xristos3490 I can see your reply from an hour ago here, but as I didn't refresh the page, I replied here without seeing your reply. 😅 Still, I believe my reply is relevant to your latest comment. 🙂

Reply:
> Sidenote: We are using name [productCollectionLocation](https://github.com/woocommerce/woocommerce/blob/53e1da8ee51d0ef76e4d13dd63cc287bd013d177/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L132) on Frontend, which contain calculated location. Therefore, we will need to update that name too in case we decide to combine these tw

Yes! 100%!

> Question: What should we call this SINGLE information? selectedReference or location?

How about a generic `currentReference`, for now? (same will be in frontend)

Reply:
Note that PHP not only needs to update the parser to reflect the new term, but also to incorporate the value of `selectedReference`. In order to achieve this, it needs access to the `selectedReference` attribute. 

This means that the PC block needs to provide that context to the product template. Does that make sense? Am I overlooking something? Is there another way to retrieve that?

Reply:
> I also thought this would be the case, but it isn't. There is one problem. [..] This function adds location context to inner blocks with query context. But, those inner blocks won't have access to selectedReference.

Oh, you are right! Also, this is probably a glitch in the 'collection' attribute 😅 🙌 

> Based on this, I think the best option is to add selectedReference inside query

Sounds like a good solution, yes. If we add the ref-related attribute in the `query`, we should probably also add the collection attritube as well 🤔

Reply:
> Oh, you are right! Also, this is probably a glitch in the 'collection' attribute

I am not sure if I understand this. Can you please explain this a bit more? 🤔

I think this is a problem only if we are trying to access 'collection' attribute in [provide_location_context_for_inner_blocks](https://github.com/woocommerce/woocommerce/blob/be8d1259631c9c80fd67ced80de7bfb4eaff4169/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L119), But I don't think we are doing that. Am I missing something? 🤔

> Sounds like a good solution, yes. If we add the ref-related attribute in the `query, ' we should probably also add the collection attribute. Tho this doesn't sound right. 🤔

Same as above, how 'collection' will be used in [provide_location_context_for_inner_blocks](https://github.com/woocommerce/woocommerce/blob/be8d1259631c9c80fd67ced80de7bfb4eaff4169/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L119)? Do we wanna add 'collection' in `productCollectionLocation` too? 🤔

Reply:
Yes, you are correct. I got confused by thinking that 3PDs would need the collection name while filtering query args, but that's all contained in the product-template block. So, yes - no issues with this attribute. 🙌 

So, the question is how and where to handle the `selectedReference` or whatever the name/structure we decide [here](https://github.com/woocommerce/woocommerce/pull/50164#discussion_r1703901015). 🤔

Reply:
Also, It doesn't feel right that 3PDs need to use `query` content and we automatically add `productCollectionLocation` to those blocks. I believe we should implement a better way. Should we consider adding a new attribute like:
- `useProductCollectionContext`
- `inheritProductCollectionContext`
- `worksWithProductCollectionContext`?

This way, 3PDs can add this attribute in `usesContext` in block.json file. The current way of doing this is working like a hack/magic. OR we can consider renaming `query` to `productCollectionQuery` to make it more clear. What do you think? 🤔

One way this would be useful is that we can deterministically know which blocks want to use Product Collection block context. Based on this, we can even add more keys to `usesContext` dynamically using hooks like `register_block_type_args`, `blocks_register_block_type_args` etc. I haven't tested it yet, but here is how it will work based on ChatGPT output:
```php
function add_custom_uses_context($args, $name) {
    // Check if it's the block you want to modify
    if ($name === 'your-namespace/your-block') {
        // If usesContext doesn't exist, initialize it
        if (!isset($args['uses_context'])) {
            $args['uses_context'] = array();
        }
        
        // Check if worksWithProductCollectionContext is already present
        if (in_array('worksWithProductCollectionContext', $args['uses_context'])) {
            // Add your additional context keys
            $args['uses_context'][] = 'yourCustomContext1';
            $args['uses_context'][] = 'yourCustomContext2';
        }
    }
    return $args;
}
add_filter('register_block_type_args', 'add_custom_uses_context', 10, 2);
```

What do you think about this? CC: @kmanijak @xristos3490

PS: This isn't directly related to this PR. I can create a separate issue for this for further discussion. 🙂

Reply:
Love this exploration, Manish! 🙌 💯 

> The current way of doing this is working like a hack/magic.

In my opinion, we ended up on this path due to the decision not to make the 'location' data an attribute. Based on the documentation:

> [!NOTE]  
> Currently, block context only supports values derived from **the block’s own attributes**. This could be enhanced in the future to support additional sources of context values.
>[Source](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-context/#providing-block-context)

The open possibilities here make me think that we could push upstream for this. A fair question would be: how is the `postId` provided in the editor without being an attribute? If we could streamline the `useGetLocation` result into a similar handling I think we have a great and consistent way to handle all this. Thoughts?

Reply:
> So, the question is how and where to handle the `selectedReference` or whatever the name/structure we decide [here](https://github.com/woocommerce/woocommerce/pull/50164#discussion_r1703901015). 🤔

@xristos3490 IMO Let's go with `currentReference` in `query` attribute. This way, we will have access to `selectedReference` in PHP.

Regarding the structure, I think we can go with current structure. If needed, we can make it an array like this in future i.e.
```tsx
[
  {
    type: 'product',
    id: 123
  },
  {
    type: 'archive',
    id: 456
  }
]
```

What do you think? 

> In my opinion, we ended up on this path due to the decision not to make the 'location' data an attribute. Based on the documentation:

Yeah, Right. However, we can improve it by creating a new attribute, as I suggested above. Right? This way, we can access various contexts in any inner block. 

> The open possibilities here make me think that we could push upstream for this. A fair question would be: how is the `postId` provided in the editor without being an attribute? If we could streamline the `useGetLocation` result into a similar handling I think we have a great and consistent way to handle all this. Thoughts?

After spending some time on this, Looks like, On the Server/PHP side, `postId` is added [here](https://github.com/WordPress/wordpress-develop/blob/b17b5d18548b7d9a1083425f2398322e0b82d613/src/wp-includes/blocks.php#L2078-L2088). Basically, `WP_Block` is created with `postId` and `postType` as already available context. By looking at [this code](https://github.com/WordPress/wordpress-develop/blob/b17b5d18548b7d9a1083425f2398322e0b82d613/src/wp-includes/blocks.php#L2110), we can say that `render_block_context` will always have access to `postId` and `postType` if it's available. Although, There is nothing magical here because it's just using `global $post` to get `postId` and `postType`. Which can be done by us as well. 

It was interesting for me to understand how `postId` is available on Frontend when it's there in `usesContext` but isn't available as an attribute. It's important because if wanna contribute upstream then it should work on both Editor and Frontend.

That being said, I am not sure if there is anything we can contribute upstream. Because we already have `render_block_context` available to us which can be used to modify the context values of any block. Also, we can use hooks like `register_block_type_args`, `blocks_register_block_type_args` etc. to add more context values to `usesContext`.

Is there anything I am missing here? Are you suggesting something else? 🤔